### PR TITLE
STM Backend support

### PR DESCRIPTION
**Pull request by psivasubramanian**
This request addresses the review comments of pull -https://github.com/Metrological/WebKitForWayland/pull/40

### CODE REVIEW

File: Source/cmake/OptionsWPE.cmake [hunk 1]
```diff
@@ -15,12 +15,16 @@ WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_CSS_GRID_LAYOUT PRIVATE ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_CSS_IMAGE_SET PUBLIC ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_CSS_REGIONS PUBLIC ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_CSS_SELECTORS_LEVEL4 PUBLIC ON)
+WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_DATABASE_PROCESS PRIVATE ON)
```
Comment (philn) [inline Source/cmake/OptionsWPE.cmake]: The changes in this file seem unrelated to the patch. Please make sure you base your branch on top of current WPE master.

Comment (psivasubramanian) [inline Source/cmake/OptionsWPE.cmake]: Addressed by rebasing files.

File: Source/WPE/CMakeLists.txt [hunk 2]
```diff
@@ -176,6 +184,22 @@ if (USE_WPE_BACKEND_WESTEROS)
...
+        ${STM_INCLUDE_DIRS}
+        ${STMEGL_INCLUDE_DIRS}
+     )
+     list(APPEND WPE_LIBRARIES
+        ${STM_LIBRARIES}
+        ${STMEGL_LIBRARIES}
+     )
+     list(APPEND WPE_SOURCES
+        Source/Graphics/STM/RenderingBackendSTM.cpp
+        Source/Graphics/RenderingBackend.cpp
```
Comment (zdobersek) [inline Source/WPE/CMakeLists.txt]: I don't think you need this line, the file is already included in the list on line 114.

File: Source/cmake/OptionsWPE.cmake [hunk 3]
```diff
@@ -95,6 +95,11 @@ if (USE_WPE_BACKEND_BCM_RPI)
     find_package(BCMHost REQUIRED)
 endif ()
 
+if (USE_WPE_BACKEND_STM)
+find_package(Wayland REQUIRED)
```
Comment (zdobersek) [inline Source/cmake/OptionsWPE.cmake]: Can use indentation.

File: Source/cmake/OptionsWPE.cmake [hunk 4]
```diff
@@ -95,6 +95,11 @@ if (USE_WPE_BACKEND_BCM_RPI)
...
+find_package(WaylandEGL REQUIRED)
```
Comment (zdobersek) [inline Source/cmake/OptionsWPE.cmake]: Can use indentation.

File: Source/WPE/Headers/WPE/ViewBackend/ViewBackend.h [hunk 5]
```diff
@@ -57,6 +60,10 @@ class ViewBackend {
     virtual void commitBuffer(int, const uint8_t*, size_t) = 0;
     virtual void destroyBuffer(uint32_t) = 0;
 
+    virtual void handleKeyboardEvent(const Input::KeyboardEvent& event);
+    virtual void handlePointerEvent(const Input::PointerEvent& event);
+    virtual void handleAxisEvent(const Input::AxisEvent& event);
```
Comment (zdobersek) [inline Source/WPE/Headers/WPE/ViewBackend/ViewBackend.h]: I don't see these three methods used, so just not add them for the time being.

File: Source/WPE/Headers/WPE/Graphics/RenderingBackend.h [hunk 6]
```diff
@@ -46,6 +52,10 @@ class RenderingBackend {
         class Client {
         public:
             virtual void destroyBuffer(uint32_t) = 0;
+/* FIXME Input handling
```
Comment (zdobersek) [inline Source/WPE/Headers/WPE/Graphics/RenderingBackend.h]: No changes to this file are effective, so you can just remove them.

File: Source/WPE/Headers/WPE/Input/Events.h [hunk 7]
```diff
@@ -29,6 +29,11 @@
 
 #include <array>
 
+namespace IPC {
```
Comment (zdobersek) [inline Source/WPE/Headers/WPE/Input/Events.h]: Same here, this IPC namespace and all the additional changes to this file are not used and shouldn't be necessary.

File: Source/WPE/Source/ViewBackend/STM/ViewBackendSTM.cpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+void ViewBackendSTM::destroyBuffer(uint32_t)
+{
+}
+
+void ViewBackendSTM::setInputClient(Input::Client* client)
+{
+    m_input_client = client;
+}
+
+void ViewBackendSTM::handleKeyboardEvent(const Input::KeyboardEvent& event)
```
Comment (zdobersek) [inline Source/WPE/Source/ViewBackend/STM/ViewBackendSTM.cpp]: While these handle*Event() methods are implemented for this class, they're never called, so don't include them at this point.

**Comment by psivasubramanian** [comment]
@zdobersek 
Incorporated all your review comments.

File: Source/WPE/Source/ViewBackend/STM/ViewBackendSTM.cpp (hunk 8 above)
Comment (aazamansari) [inline Source/WPE/Source/ViewBackend/STM/ViewBackendSTM.cpp]: Once the new IPC mechanism is available, we should call frame complete by registering a callback with wayland for frame complete event.

**Comment by zdobersek** [comment]
Looks OK, thanks for the updates.

**Comment by psivasubramanian** [comment]
Thanks @zdobersek

**Comment by armcc** [comment]
This breaks non-Wayland nexus builds:

`| In file included from /.../0.1+gitAUTOINC+30f856124f-r0/git/Source/WPE/Source/Graphics/RenderingBackend.cpp:32:0:
| /.../0.1+gitAUTOINC+30f856124f-r0/git/Source/WPE/Source/Graphics/STM/RenderingBackendSTM.h:5:25: fatal error: wayland-egl.h: No such file or directory
| compilation terminated.
`

Shouldn't inclusion of the header files for each backend be conditional on the associated backend being enabled?

**Comment by albertd** [comment]
@armcc added a guard in the header file in master

**Comment by psivasubramanian** [comment]
Thanks @albertd

**Comment by blino** [comment]
See #79 for the port on new WPE
